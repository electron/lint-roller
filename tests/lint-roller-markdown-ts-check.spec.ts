import * as cp from 'node:child_process';
import * as path from 'node:path';

const FIXTURES_DIR = path.resolve(__dirname, 'fixtures');

function runLintMarkdownTsCheck(...args: string[]) {
  return cp.spawnSync(
    process.execPath,
    [path.resolve(__dirname, '../dist/bin/lint-markdown-ts-check.js'), ...args],
    {
      stdio: 'pipe',
      encoding: 'utf-8',
      cwd: FIXTURES_DIR,
      env: { NODE_OPTIONS: '--no-deprecation' },
    },
  );
}

describe('lint-roller-markdown-ts-check', () => {
  it('should run clean when there are no errors', () => {
    const { status, stdout } = runLintMarkdownTsCheck('--root', FIXTURES_DIR, 'ts-check-clean.md');

    expect(stdout).toEqual('');
    expect(status).toEqual(0);
  });

  it('can ignore a glob', () => {
    const { status } = runLintMarkdownTsCheck(
      '--root',
      FIXTURES_DIR,
      '--ignore',
      '**/ts-check.md',
      '{ts-check,ts-check-clean}.md',
    );

    expect(status).toEqual(0);
  });

  it('can ignore multiple globs', () => {
    const { status } = runLintMarkdownTsCheck(
      '--root',
      FIXTURES_DIR,
      '--ignore',
      '**/ts-check.md',
      '--ignore',
      '**/dirty.md',
      '{dirty,ts-check,ts-check-clean}.md',
    );

    expect(status).toEqual(0);
  });

  it('can ignore from a file', () => {
    const { status } = runLintMarkdownTsCheck(
      '--root',
      FIXTURES_DIR,
      '--ignore-path',
      path.resolve(FIXTURES_DIR, 'ignorepaths'),
      '{dirty,ts-check,ts-check-clean}.md',
    );

    expect(status).toEqual(0);
  });

  it('should type check code blocks', () => {
    const { status, stdout } = runLintMarkdownTsCheck('--root', FIXTURES_DIR, 'ts-check.md');

    expect(stdout.replace(FIXTURES_DIR, '<root>')).toMatchSnapshot();
    expect(status).toEqual(1);
  });

  it('can use a custom config', () => {
    const { status, stdout, stderr } = runLintMarkdownTsCheck(
      '--root',
      FIXTURES_DIR,
      '--config',
      'custom-config.json',
      'ts-check-custom-config.md',
    );

    expect(stdout).toEqual('');
    expect(stderr).toEqual('');
    expect(status).toEqual(0);
  });
});
