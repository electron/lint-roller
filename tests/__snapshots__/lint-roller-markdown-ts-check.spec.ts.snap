// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lint-roller-markdown-ts-check should type check code blocks 1`] = `
"ts-check.md:37:1: Code block has both @ts-nocheck and @ts-expect-error/@ts-type/@ts-window-type, they conflict
ts-check.md:43:1: Code block has both @ts-nocheck and @ts-expect-error/@ts-type/@ts-window-type, they conflict
[96mts-check.md[0m:[93m109[0m:[93m5[0m - [91merror[0m[90m TS2304: [0mCannot find name 'a'.

[7m109[0m if (a > b) {
[7m   [0m [91m    ~[0m

[96mts-check.md[0m:[93m109[0m:[93m9[0m - [91merror[0m[90m TS2304: [0mCannot find name 'b'.

[7m109[0m if (a > b) {
[7m   [0m [91m        ~[0m

[96mts-check.md[0m:[93m112[0m:[93m28[0m - [91merror[0m[90m TS2304: [0mCannot find name 'a'.

[7m112[0m   console.log(\`not true: \${a} < \${b}\`)
[7m   [0m [91m                           ~[0m

[96mts-check.md[0m:[93m112[0m:[93m35[0m - [91merror[0m[90m TS2304: [0mCannot find name 'b'.

[7m112[0m   console.log(\`not true: \${a} < \${b}\`)
[7m   [0m [91m                                  ~[0m

[96mts-check.md[0m:[93m115[0m:[93m8[0m - [91merror[0m[90m TS2339: [0mProperty 'AwesomeAPI' does not exist on type 'Window & typeof globalThis'.

[7m115[0m window.AwesomeAPI.bar('baz')
[7m   [0m [91m       ~~~~~~~~~~[0m

[96mts-check.md[0m:[93m129[0m:[93m15[0m - [91merror[0m[90m TS2339: [0mProperty 'wrongAPI' does not exist on type 'typeof BrowserWindow'.

[7m129[0m BrowserWindow.wrongAPI('foo')
[7m   [0m [91m              ~~~~~~~~[0m

[96mts-check.md[0m:[93m135[0m:[93m15[0m - [91merror[0m[90m TS2339: [0mProperty 'wrongAPI' does not exist on type 'typeof BrowserWindow'.

[7m135[0m BrowserWindow.wrongAPI('foo')
[7m   [0m [91m              ~~~~~~~~[0m

[96mts-check.md[0m:[93m167[0m:[93m8[0m - [91merror[0m[90m TS2339: [0mProperty 'AwesomeAPI' does not exist on type 'Window & typeof globalThis'.

[7m167[0m window.AwesomeAPI.foo(42)
[7m   [0m [91m       ~~~~~~~~~~[0m

[96mts-check.md[0m:[93m4[0m:[93m9[0m - [91merror[0m[90m TS2339: [0mProperty 'foo' does not exist on type 'Console'.

[7m4[0m console.foo('whoops')
[7m [0m [91m        ~~~[0m

[96mts-check.md[0m:[93m54[0m:[93m15[0m - [91merror[0m[90m TS2339: [0mProperty 'wrongAPI' does not exist on type 'typeof BrowserWindow'.

[7m54[0m BrowserWindow.wrongAPI('foo')
[7m  [0m [91m              ~~~~~~~~[0m

[96mts-check.md[0m:[93m60[0m:[93m15[0m - [91merror[0m[90m TS2339: [0mProperty 'wrongAPI' does not exist on type 'typeof BrowserWindow'.

[7m60[0m BrowserWindow.wrongAPI('foo')
[7m  [0m [91m              ~~~~~~~~[0m

[96mts-check.md[0m:[93m8[0m:[93m9[0m - [91merror[0m[90m TS2339: [0mProperty 'foo' does not exist on type 'Console'.

[7m8[0m console.foo('whoops')
[7m [0m [91m        ~~~[0m

[96mts-check.md[0m:[93m83[0m:[93m8[0m - [91merror[0m[90m TS2339: [0mProperty 'myAwesomeAPI' does not exist on type 'Window & typeof globalThis'.

[7m83[0m window.myAwesomeAPI()
[7m  [0m [91m       ~~~~~~~~~~~~[0m


Found 13 errors in 9 files.

Errors  Files
     5  ts-check.md[90m:109[0m
     1  ts-check.md[90m:129[0m
     1  ts-check.md[90m:135[0m
     1  ts-check.md[90m:167[0m
     1  ts-check.md[90m:4[0m
     1  ts-check.md[90m:54[0m
     1  ts-check.md[90m:60[0m
     1  ts-check.md[90m:8[0m
     1  ts-check.md[90m:83[0m

"
`;
